module/os/linux/zfs/zvol_os.c

```diff
@@ -734,6 +740,13 @@ zvol_open(struct block_device *bdev, fmode_t flag)
 	}
 
 	mutex_enter(&zv->zv_state_lock);
+
+	if (unlikely(zv->zv_flags & ZVOL_REMOVING)) {
+		mutex_exit(&zv->zv_state_lock);
+		rw_exit(&zvol_state_lock);
+		return (-SET_ERROR(ENXIO));
+	}
+
```

Comment:
nit: could you simply this block by re-arranging it to?:
```C
	if (unlikely(zv->zv_flags & ZVOL_REMOVING)) {
		rw_exit(&zvol_state_lock);
		return (-SET_ERROR(ENXIO));
	}

	mutex_enter(&zv->zv_state_lock);
```

Reply:
I'd need to think it through. The obvious concern is that by not taking the lock is that it could miss the transition to `ZVOL_REMOVING`, and end up taking another open count when another thread is expecting the count to only go down. But that might be nothing, like I say, gotta think about it.

It's probably safer just to take the lock. It's not performance-critical.

Reply:
Thanks, that makes sense